mesonbuild/interpreter/interpreter.py

```diff
@@ -3382,3 +3387,53 @@ def func_range(self, node, args: T.Tuple[int, T.Optional[int], T.Optional[int]],
         if step < 1:
             raise InterpreterException('step must be >=1')
         return P_OBJ.RangeHolder(start, stop, step, subproject=self.subproject)
+
+    @FeatureNew('response_file', '1.2.0')
+    @typed_pos_args('response_file', varargs=(str, mesonlib.File))
+    @typed_kwargs(
+        'response_file',
+        KwargInfo('prefix', str, default='@'),
+        KwargInfo('quote', (bool, str), default=True, validator=in_set_validator({'sh', 'gcc', 'cmd', True, False})),
+        KwargInfo('max_args_length', (NoneType, int)),
+        KwargInfo('separator', str, default=' '),
+        KwargInfo('relative_paths', bool, default=False),
+    )
+    def func_response_file(self, node: mparser.BaseNode, args: T.Tuple[T.List[mesonlib.FileOrString]], kwargs: kwtypes.ResponseFile) -> T.List[str]:
+        if not args[0]:
+            return []
+
+        max_args_length = kwargs['max_args_length']
+        if max_args_length is None:
+            max_args_length = get_rsp_threshold()
+
+        quote_fcts = {
+            'sh': shlex.quote,
+            'gcc': gcc_rsp_quote,
+            'cmd': cmd_quote,
+            True: mesonlib.quote_arg,
+        }
+        quote_style = kwargs['quote']
+
+        def quote(value: mesonlib.FileOrString):
+            if isinstance(value, mesonlib.File):
+                if kwargs['relative_paths']:
+                    value = value.relative_name()
+                else:
+                    value = value.absolute_path(self.environment.source_dir, self.environment.build_dir)
+            if quote_style:
+                value = quote_fcts[quote_style](value)
+            return value
+
+        quoted_args = kwargs['separator'].join(quote(a) for a in args[0])
+
+        if len(quoted_args) <= max_args_length:
+            mlog.debug(f'Estimated arg length {len(quoted_args)} <= {max_args_length}')
+            return args[0]
+
+        rsp_index = next(self.rsp_files[self.subdir])
+        rsp_file = Path(self.subdir) / f'args{rsp_index}.rsp'
+
+        mlog.debug(f'Writing rsp file: {rsp_file}')
+        Path(self.environment.build_dir, rsp_file).write_text(quoted_args, encoding='utf-8')
```

Comment:
I realize there is a big problem with this rsp file: If a file changes it won't trigger reconfigure or rebuild.

Returning a list[str] is probably not good enough. We need to return a new type of object that also contains all depend_files. If that object gets used in `run_command()` those depenf_files will have to be added using `interpreter.add_build_def_file()`. If used in a custom_target() they have to be added to that target's depends.

Reply:
You're right! I put this PR in *draft* until I address it.